pkg/controllers/nodepool/readiness/controller.go

```diff
@@ -0,0 +1,130 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodePool *v1beta1.NodePool) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodepool.status").With("nodepool", nodePool.Name))
+	ctx = injection.WithControllerName(ctx, "nodepool.status")
+	stored := nodePool.DeepCopy()
+	supportedNC := c.cloudProvider.GetSupportedNodeClasses()
+	var errs error
+	var ready bool
+	nodeClass, err := c.hasNodeClass(ctx, supportedNC, nodePool)
+	if err != nil {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "UnresolvedNodeClass", "Unable to resolve nodeClass")
+		errs = multierr.Append(errs, err)
+	} else {
+		ready = c.setReadyCondition(nodePool, nodeClass)
+	}
+	if !equality.Semantic.DeepEqual(stored, nodePool) {
+		if err := c.kubeClient.Status().Patch(ctx, nodePool, client.MergeFrom(stored)); err != nil {
+			if errors.IsConflict(err) {
+				return reconcile.Result{Requeue: true}, nil
+			}
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+	if errs != nil {
+		return reconcile.Result{}, errs
+	}
+	return reconcile.Result{Requeue: !ready}, nil
+}
+
+func (c *Controller) setReadyCondition(nodePool *v1beta1.NodePool, nodeClass status.Object) bool {
+	ready := nodeClass.StatusConditions().Get(status.ConditionReady)
+	if ready.IsUnknown() {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "NodeClassNotReady", "Node Class Not Ready")
+		return false
+	}
+	if ready.IsFalse() {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, ready.Reason, ready.Message)
+		return false
+	}
+	nodePool.StatusConditions().SetTrue(v1beta1.ConditionTypeNodeClassReady)
+	return true
+}
+
+func (c *Controller) hasNodeClass(ctx context.Context, supportedNC []status.Object, nodePool *v1beta1.NodePool) (status.Object, error) {
+	if len(supportedNC) == 0 {
+		return nil, fmt.Errorf("resolving nodeClass")
+	}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodePool.Spec.Template.Spec.NodeClassRef.Name}, supportedNC[0]); err != nil {
+		return nil, fmt.Errorf("resolving nodeClass, %w", err)
+	}
+	return supportedNC[0], nil
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	builder := controllerruntime.NewControllerManagedBy(m)
+	for _, supportedNC := range c.cloudProvider.GetSupportedNodeClasses() {
```

Comment:
Like with the other controller, I don't think that we need to use unstructured here. I think we can just directly use the types